build-support/bin/native/bootstrap_code.sh

```diff
@@ -0,0 +1,106 @@
+REPO_ROOT=$(cd $(dirname "${BASH_SOURCE[0]}") && cd ../../.. && pwd -P)
+
+# Defines:
+# + CACHE_ROOT: The pants cache directory, ie: ~/.cache/pants.
+# Exposes:
+# + die: Exit in a failure state and optionally log an error message to the console.
+# + fingerprint_data: Fingerprints the data on stdin.
+source ${REPO_ROOT}/build-support/common.sh
+
+source ${REPO_ROOT}/build-support/bin/native/bootstrap_rust.sh
+
+readonly KERNEL=$(uname -s | tr '[:upper:]' '[:lower:]')
+case "${KERNEL}" in
+  linux)
+    readonly LIB_EXTENSION=so
+    ;;
+  darwin)
+    readonly LIB_EXTENSION=dylib
+    ;;
+  *)
+    die "Unknown kernel ${KERNEL}, cannot bootstrap pants native code!"
+    ;;
+esac
+
+readonly NATIVE_ROOT="${REPO_ROOT}/src/rust/engine"
+readonly NATIVE_ENGINE_BINARY="native_engine.so"
+readonly NATIVE_ENGINE_RESOURCE="${REPO_ROOT}/src/python/pants/engine/${NATIVE_ENGINE_BINARY}"
+
+# N.B. Set $MODE to "debug" to generate a binary with debugging symbols.
+readonly MODE="${MODE:-release}"
+case "$MODE" in
+  debug) MODE_FLAG="" ;;
+  *) MODE_FLAG="--release" ;;
+esac
+
+readonly NATIVE_ENGINE_CACHE_DIR=${CACHE_ROOT}/bin/native-engine
+
+function calculate_current_hash() {
+  # Cached and unstaged files, with ignored files excluded.
+  # NB: We fork a subshell because one or both of `ls-files`/`hash-object` are
+  # sensitive to the CWD, and the `--work-tree` option doesn't seem to resolve that.
+  (
+   cd ${REPO_ROOT}
+   (echo "${MODE_FLAG}"
+    echo "${RUST_TOOLCHAIN}"
+    git ls-files -c -o --exclude-standard \
+     "${NATIVE_ROOT}" \
+     "${REPO_ROOT}/src/python/pants/engine/native.py" \
+     "${REPO_ROOT}/build-support/bin/native" \
+     "${REPO_ROOT}/3rdparty/python/requirements.txt" \
+   | grep -v -E -e "/BUILD$" -e "/[^/]*\.md$" \
+   | git hash-object -t blob --stdin-paths) | fingerprint_data
+  )
+}
+
+function _build_native_code() {
+  # Builds the native code, and echos the path of the built binary.
+
+  "${REPO_ROOT}/build-support/bin/native/cargo" build ${MODE_FLAG} \
```

Comment:
Nit: Missing a few quotes on this and the next line

Reply:
I don't think MODE_FLAG should be quoted, flags might be in `-x y` form, but fixed the `--manifest-path` argument.

Reply:
I can get on board with that, but the `--manifest-path` arg doesn't look fixed :)